File: source/native/src/expression/Value.cpp
#include <cmath>
#include <cerrno>
#include <string>
#include <String.hpp>
#include "Value.hpp"

Value::Value(): type_(Value::TYPE_BOTTOM)
{
}

Value::Value(int value): type_(Value::TYPE_INT), intValue_(value)
{
}

Value::Value(bool value): type_(Value::TYPE_INT), intValue_(value)
{
}

Value::Value(float value): type_(Value::TYPE_FLOAT), floatValue_(value)
{
}

Value::Value(string value): type_(Value::TYPE_STRING), stringValue_(value)
{
}

Value::Value(Variant variant)
{
	if (variant.get_type() == Variant::BOOL) {
        type_ = Value::TYPE_INT;
        intValue_ = variant ? 1 : 0;
    } else if (variant.get_type() == Variant::INT) {
        type_ = Value::TYPE_INT;
        intValue_ = (int) variant;
    } else if (variant.get_type() == Variant::REAL) {
        type_ = Value::TYPE_FLOAT;
        floatValue_ = (float) variant;
    } else if (variant.get_type() == Variant::STRING) {
        type_ = Value::TYPE_STRING;
        String stringValue = variant;
        stringValue_ = string(stringValue.utf8().get_data());
    } else if (variant.get_type() == Variant::ARRAY) {
        type_ = Value::TYPE_ARRAY;
        vector<Value> values_;
        Array array_ = variant;
        for (int i = 0; i < array_.size(); i++) {
            values_.push_back(array_[i]);
        }
        arrayValue_ = values_;
    } else {
        type_ = Value::TYPE_BOTTOM;
    }
}

Value::Value(vector<Value> values_): type_(Value::TYPE_ARRAY), arrayValue_(values_)
{
}

int Value::type() const
{
    return type_;
}

bool Value::isInt() const
{
    return type_ == Value::TYPE_INT;
}

bool Value::isFloat() const
{
    return type_ == Value::TYPE_FLOAT;
}

bool Value::isString() const
{
    return type_ == Value::TYPE_STRING;
}

bool Value::isArray() const
{
    return type_ == Value::TYPE_ARRAY;
}

bool Value::isArithmetic() const
{
    return isInt() || isFloat();
}

bool Value::isComparable() const
{
    return isArithmetic() || isString();
}

bool Value::isBottom() const
{
    return type_ == Value::TYPE_BOTTOM;
}

int Value::intValue() const
{
    if (type_ == Value::TYPE_INT) {
        return intValue_;
    }

    if (type_ == Value::TYPE_FLOAT) {
        return (int) floatValue_;
    }

    return 0;
}

float Value::floatValue() const
{
    if (type_ == Value::TYPE_INT) {
        return (float) intValue_;
    }

    if (type_ == Value::TYPE_FLOAT) {
        return floatValue_;
    }

    return 0.0;
}

string Value::stringValue() const
{
    if (type_ == Value::TYPE_INT) {
        return to_string(intValue_);
    }

    if (type_ == Value::TYPE_FLOAT) {
        return to_string(floatValue_);
    }

    if (type_ == Value::TYPE_STRING) {
        return stringValue_;
    }

    return "";
}

Value Value::equal(const Value &other) const
{
    if (!isComparable() || !other.isComparable()) {
        return Value();
    }

    if (isString() || other.isString()) {
        return Value(stringValue() == other.stringValue());
    }

    if (isInt() && other.isInt()) {
        return Value(intValue() == other.intValue());
    }

    return Value(floatValue() == other.floatValue());
}

Value Value::logicalNot() const
{
    if (!isArithmetic()) {
        return Value();
    }

    return Value(0 == intValue());
}

Value Value::logicalAnd(const Value &other) const
{
    if (!isArithmetic() || !other.isArithmetic()) {
        return Value();
    }

    return Value(intValue() && other.intValue());
}

Value Value::logicalOr(const Value &other) const
{
    if (!isArithmetic() || !other.isArithmetic()) {
        return Value();
    }

    return Value(intValue() || other.intValue());
}

Value Value::logicalXor(const Value &other) const
{
    if (!isArithmetic() || !other.isArithmetic()) {
        return Value();
    }

    return Value(intValue() != other.intValue());
}

Value Value::bitwiseNot() const
{
    if (isArithmetic() || !isArithmetic()) {
        return Value();
    }

    return Value(~intValue());
}

Value Value::bitwiseAnd(const Value &other) const
{
    if (!isArithmetic() || !other.isArithmetic() || !isInt() || !other.isInt()) {
        return Value();
    }

    return Value(intValue() & other.intValue());
}

Value Value::bitwiseOr(const Value &other) const
{
    if (!isArithmetic() || !other.isArithmetic() || !isInt() || !other.isInt()) {
        return Value();
    }

    return Value(intValue() | other.intValue());
}

Value Value::bitwiseXor(const Value &other) const
{
    if (!isArithmetic() || !other.isArithmetic() || !isInt() || !other.isInt()) {
        return Value();
    }

    return Value(intValue() ^ other.intValue());
}

Value Value::compare(const Value &other) const
{
    if (!isComparable() || !other.isComparable()) {
        return Value();
    }

    if (isString() || other.isString()) {
        if (stringValue() == other.stringValue()) {
            return Value(0);
        }

        return Value(stringValue().size() < other.stringValue().size() ? -1 : 1);
    }

    if (isInt() && other.isInt()) {
        if (intValue() == other.intValue()) {
            return Value(0);
        }

        if (intValue() < other.intValue()) {
            return Value(-1);
        }

        return Value(1);
    }

    if (floatValue() == other.floatValue()) {
        return Value(0);
    }

    if (floatValue() < other.floatValue()) {
        return Value(-1);
    }

    return Value(1);
}

Value Value::greater(const Value &value) const
{
    Value diff = compare(value);

    if (diff.isBottom()) {
        return diff;
    }

    return Value(diff.intValue() > 0);
}

Value Value::greaterOrEqual(const Value &value) const
{
    Value diff = compare(value);

    if (diff.isBottom()) {
        return diff;
    }

    return Value(diff.intValue() >= 0);
}

Value Value::less(const Value &value) const
{
    Value diff = compare(value);

    if (diff.isBottom()) {
        return diff;
    }

    return Value(diff.intValue() < 0);
}

Value Value::lessOrEqual(const Value &value) const
{
    Value diff = compare(value);

    if (diff.isBottom()) {
        return diff;
    }

    return Value(diff.intValue() <= 0);
}

Value Value::add(const Value &other) const
{
    if (isString() || other.isString()) {
        return stringValue() + other.stringValue();
    }

    if (!isArithmetic() || !other.isArithmetic()) {
        return Value();
    }

    if (isInt() && other.isInt()) {
        return Value(intValue() + other.intValue());
    }

    return Value(floatValue() + other.floatValue());
}


Value Value::subtract(const Value &other) const
{
    if (!isArithmetic() || !other.isArithmetic()) {
        return Value();
    }

    if (isString() || other.isString()) {
        return Value(); // TODO: Review string behavior
    }

    if (isInt() && other.isInt()) {
        return Value(intValue() - other.intValue());
    }

    return Value(floatValue() - other.floatValue());
}

Value Value::inverse() const
{
    if (!isArithmetic()) {
        return Value();
    }

    if (isInt()) {
        return Value(-intValue());
    }

    return Value(-floatValue());
}

Value Value::mod(const Value &other) const
{
    if (!isArithmetic() || !other.isArithmetic() || isFloat() || other.isFloat() || other.equal(Value(0))) {
        return Value();
    }

    return Value(intValue() % other.intValue());
}

Value Value::multiply(const Value &other) const
{
    if (!isArithmetic() || !other.isArithmetic()) {
        return Value();
    }

    if (isInt() && other.isInt()) {
        return Value(intValue() * other.intValue());
    }

    return Value(floatValue() * other.floatValue());
}

Value Value::divide(const Value &other) const
{
    if (!isArithmetic() || !other.isArithmetic() || other.equal(Value(0))) {
        return Value();
    }

    if (isInt() && other.isInt()) {
        return Value(intValue() / other.intValue());
    }

    return Value(floatValue() / other.floatValue());
}

Value Value::pow(const Value &other) const
{
    if (!isArithmetic() || !other.isArithmetic()) {
        return Value();
    }

    // TODO: Return bottom on invalid pow

    if (isInt() && other.isInt()) {
        return Value((int) std::pow((int) intValue(), (int) other.intValue()));
    }

    return Value((float) std::pow((int) floatValue(), (int) other.floatValue()));
}

Value::operator bool() const
{
    if (!isArithmetic()) {
        return false;
    }

    if (isInt()) {
        return intValue() != 0;
    }

    if (isString()) {
        return stringValue().size() > 0;
    }

    return floatValue() != 0;
}

Value::operator int() const
{
    if (!isArithmetic()) {
        return 0;
    }

    return intValue();
}

Value::operator float() const
{
    if (!isArithmetic()) {
        return 0;
    }

    return floatValue();
}

Value::operator Variant() const
{
    if (isInt()) {
        return Variant(intValue());
    }

    if (isFloat()) {
        return Variant(floatValue());
    }

    if (isString()) {
        return Variant(stringValue().c_str());
    }

    if (isArray()) {
        Array array_;
        for (int i = 0; i < arrayValue_.size(); i++) {
            array_.push_back(arrayValue_[i]);
        }
        return Variant(array_);
    }

    return Variant();
}

File: source/native/src/expressions/value.rs
use gdnative::prelude::*;
use std::fmt;

pub enum Value {
    IntValue(i64),
    FloatValue(f64),
    StringValue(String),
    ArrayValue(Vec<Value>),
    BottomValue,
}

impl Clone for Value {
    fn clone(&self) -> Value {
        match self {
            Value::IntValue(x) => Value::IntValue(*x),
            Value::FloatValue(x) => Value::FloatValue(*x),
            Value::StringValue(x) => Value::StringValue(x.to_string()),
            Value::ArrayValue(x) => Value::ArrayValue(x.to_vec()),
            Value::BottomValue => Value::BottomValue,
        }
    }
}

impl From<Variant> for Value {
    fn from(variant: Variant) -> Self {
        match variant.get_type() {
            VariantType::Bool => Value::IntValue(if bool::from_variant(&variant).unwrap() {
                1
            } else {
                0
            }),
            VariantType::I64 => Value::IntValue(i64::from_variant(&variant).unwrap()),
            VariantType::F64 => Value::FloatValue(f64::from_variant(&variant).unwrap()),
            VariantType::GodotString => Value::StringValue(String::from_variant(&variant).unwrap()),
            VariantType::VariantArray => {
                let variant_array = VariantArray::from_variant(&variant).unwrap();
                let mut vec: Vec<Value> = Vec::new();
                for item in variant_array.iter() {
                    vec.push(item.into());
                }
                Value::ArrayValue(vec)
            }
            _ => Value::BottomValue,
        }
    }
}

impl ToVariant for Value {
    fn to_variant(&self) -> Variant {
        match self {
            Value::IntValue(x) => Variant::new(*x),
            Value::FloatValue(x) => Variant::new(*x),
            Value::StringValue(x) => Variant::new(x),
            Value::ArrayValue(x) => {
                let array = VariantArray::new();
                for val in x.iter() {
                    array.push(&val.to_variant());
                }
                array.into_shared().to_variant()
            }
            _ => Variant::nil(),
        }
    }
}

impl fmt::Display for Value {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        let text = match self {
            Value::FloatValue(x) => x.to_string(),
            Value::IntValue(x) => x.to_string(),
            Value::StringValue(x) => x.to_string(),
            _ => "bottom".to_string(),
        };

        write!(f, "{}", text)
    }
}

impl Value {
    pub fn to_i64(&self) -> i64 {
        match self {
            Value::FloatValue(x) => *x as i64,
            Value::IntValue(x) => *x,
            Value::StringValue(_) => 0,
            _ => 0,
        }
    }

    pub fn to_f64(&self) -> f64 {
        match self {
            Value::IntValue(x) => *x as f64,
            Value::FloatValue(x) => *x,
            Value::StringValue(_) => 0.0,
            _ => 0.0,
        }
    }

    pub fn is_string(&self) -> bool {
        matches!(self, Value::StringValue(_))
    }

    pub fn to_bool(&self) -> bool {
        match self {
            Value::FloatValue(x) => *x != 0.0,
            Value::IntValue(x) => *x != 0,
            Value::StringValue(x) => !x.is_empty(),
            Value::ArrayValue(x) => !x.is_empty(),
            _ => false,
        }
    }

    pub fn is_int(&self) -> bool {
        matches!(self, Value::IntValue(_))
    }

    pub fn is_float(&self) -> bool {
        matches!(self, Value::FloatValue(_))
    }

    pub fn is_bottom(&self) -> bool {
        matches!(self, Value::BottomValue)
    }

    pub fn is_arithmetic(&self) -> bool {
        self.is_int() || self.is_float()
    }

    pub fn is_comparable(&self) -> bool {
        self.is_arithmetic() || self.is_string()
    }

    pub fn equal(&self, other: &Value) -> Value {
        if !self.is_comparable() {
            return Value::BottomValue;
        }

        if self.is_string() || other.is_string() {
            return Value::IntValue(if self.to_string() == other.to_string() {
                1
            } else {
                0
            });
        }

        if self.is_int() && other.is_int() {
            return Value::IntValue(if self.to_i64() == other.to_i64() {
                1
            } else {
                0
            });
        }

        if self.is_float() || other.is_float() {
            return Value::IntValue(if (self.to_f64() - other.to_f64()).abs() < f64::EPSILON {
                1
            } else {
                0
            });
        }

        Value::BottomValue
    }

    pub fn logical_not(&self) -> Value {
        if !self.is_arithmetic() {
            return Value::BottomValue;
        }

        Value::IntValue(if 0 == self.to_i64() { 1 } else { 0 })
    }

    pub fn logical_and(&self, other: &Value) -> Value {
        if !self.is_arithmetic() || !other.is_arithmetic() {
            return Value::BottomValue;
        }

        Value::IntValue(if self.to_bool() && other.to_bool() {
            1
        } else {
            0
        })
    }

    pub fn logical_or(&self, other: &Value) -> Value {
        if !self.is_arithmetic() || !other.is_arithmetic() {
            return Value::BottomValue;
        }

        Value::IntValue(if self.to_bool() || other.to_bool() {
            1
        } else {
            0
        })
    }

    pub fn logical_xor(&self, other: &Value) -> Value {
        if !self.is_arithmetic() || !other.is_arithmetic() {
            return Value::BottomValue;
        }

        Value::IntValue(if self.to_bool() != other.to_bool() {
            1
        } else {
            0
        })
    }

    pub fn bitwise_not(&self) -> Value {
        if self.is_arithmetic() {
            return Value::BottomValue;
        }

        Value::IntValue(!self.to_i64())
    }

    pub fn bitwise_and(&self, other: &Value) -> Value {
        if !self.is_arithmetic() || !other.is_arithmetic() || !self.is_int() || !other.is_int() {
            return Value::BottomValue;
        }

        Value::IntValue(self.to_i64() & other.to_i64())
    }

    pub fn bitwise_or(&self, other: &Value) -> Value {
        if !self.is_arithmetic() || !other.is_arithmetic() || !self.is_int() || !other.is_int() {
            return Value::BottomValue;
        }

        Value::IntValue(self.to_i64() | other.to_i64())
    }

    pub fn bitwise_xor(&self, other: &Value) -> Value {
        if !self.is_arithmetic() || !other.is_arithmetic() || !self.is_int() || !other.is_int() {
            return Value::BottomValue;
        }

        Value::IntValue(self.to_i64() ^ other.to_i64())
    }

    pub fn compare(&self, other: &Value) -> Value {
        if !self.is_comparable() || !other.is_comparable() {
            return Value::BottomValue;
        }

        if self.is_string() || other.is_string() {
            if self.to_string() == other.to_string() {
                return Value::IntValue(0);
            }

            return Value::IntValue(if self.to_string().len() < other.to_string().len() {
                -1
            } else {
                1
            });
        }

        if self.is_int() && other.is_int() {
            if self.to_i64() == other.to_i64() {
                return Value::IntValue(0);
            }

            if self.to_i64() < other.to_i64() {
                return Value::IntValue(-1);
            }

            return Value::IntValue(1);
        }

        if (self.to_f64() - other.to_f64()).abs() < f64::EPSILON {
            return Value::IntValue(0);
        }

        if self.to_f64() < other.to_f64() {
            return Value::IntValue(-1);
        }

        Value::IntValue(1)
    }

    pub fn greater(&self, other: &Value) -> Value {
        let diff = self.compare(other);

        if diff.is_bottom() {
            return diff;
        }

        Value::IntValue(if diff.to_i64() > 0 { 1 } else { 0 })
    }

    pub fn greater_or_equal(&self, other: &Value) -> Value {
        let diff = self.compare(other);

        if diff.is_bottom() {
            return diff;
        }

        Value::IntValue(if diff.to_i64() >= 0 { 1 } else { 0 })
    }

    pub fn less(&self, other: &Value) -> Value {
        let diff = self.compare(other);

        if diff.is_bottom() {
            return diff;
        }

        Value::IntValue(if diff.to_i64() < 0 { 1 } else { 0 })
    }

    pub fn less_or_equal(&self, other: &Value) -> Value {
        let diff = self.compare(other);

        if diff.is_bottom() {
            return diff;
        }

        Value::IntValue(if diff.to_i64() <= 0 { 1 } else { 0 })
    }

    pub fn add(&self, other: &Value) -> Value {
        if self.is_string() || other.is_string() {
            return Value::StringValue(self.to_string() + &other.to_string());
        }

        if !self.is_arithmetic() || !other.is_arithmetic() {
            return Value::BottomValue;
        }

        if self.is_int() && other.is_int() {
            return Value::IntValue(self.to_i64() + other.to_i64());
        }

        Value::FloatValue(self.to_f64() + other.to_f64())
    }

    pub fn subtract(&self, other: &Value) -> Value {
        if !self.is_arithmetic() || !other.is_arithmetic() {
            return Value::BottomValue;
        }

        if self.is_int() && other.is_int() {
            return Value::IntValue(self.to_i64() - other.to_i64());
        }

        Value::FloatValue(self.to_f64() - other.to_f64())
    }

    pub fn inverse(&self) -> Value {
        if !self.is_arithmetic() {
            return Value::BottomValue;
        }

        if self.is_int() {
            return Value::IntValue(-self.to_i64());
        }

        Value::FloatValue(-self.to_f64())
    }

    pub fn modl(&self, other: &Value) -> Value {
        if !self.is_arithmetic()
            || !other.is_arithmetic()
            || self.is_float()
            || other.is_float()
            || other.equal(&Value::IntValue(0)).to_bool()
        {
            return Value::BottomValue;
        }

        Value::IntValue(self.to_i64() % other.to_i64())
    }

    pub fn multiply(&self, other: &Value) -> Value {
        if !self.is_arithmetic() || !other.is_arithmetic() {
            return Value::BottomValue;
        }

        if self.is_int() && other.is_int() {
            return Value::IntValue(self.to_i64() * other.to_i64());
        }

        Value::FloatValue(self.to_f64() * other.to_f64())
    }

    pub fn divide(&self, other: &Value) -> Value {
        if !self.is_arithmetic() || !other.is_arithmetic() {
            return Value::BottomValue;
        }

        if self.is_int() && other.is_int() {
            return Value::IntValue(self.to_i64() / other.to_i64());
        }

        Value::FloatValue(self.to_f64() / other.to_f64())
    }

    pub fn pow(&self, other: &Value) -> Value {
        if !self.is_arithmetic() || !other.is_arithmetic() {
            return Value::BottomValue;
        }

        if self.is_int() && other.is_int() {
            return Value::IntValue(self.to_i64().pow(other.to_i64() as u32));
        }

        Value::FloatValue(self.to_f64().powf(other.to_f64()))
    }
}
